File: Tests/ClipboardDrawerTests/ClipRepositoryTests.swift
@testable import ClipboardDrawer
import XCTest

final class ClipRepositoryTests: XCTestCase {
    func testUpdatePlainTextChangesEditableTextClipOnly() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: directory) }

        let repository = ClipRepository(databaseURL: directory.appendingPathComponent("clips.sqlite"))
        let textClip = ClipItem(
            id: "text",
            createdAt: Date(timeIntervalSince1970: 1),
            type: .text,
            plainText: "before",
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "before-hash",
            origin: .original,
            derivedFromClipID: nil
        )

        try repository.insert(textClip)
        try repository.updatePlainText(id: textClip.id, text: "after", contentHash: "after-hash")

        let updatedClip = try XCTUnwrap(repository.fetch(limit: 1).first)
        XCTAssertEqual(updatedClip.plainText, "after")
        XCTAssertEqual(updatedClip.contentHash, "after-hash")
        XCTAssertEqual(updatedClip.createdAt, textClip.createdAt)
    }

    func testPruneKeepsNewestAndReportsPayloads() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: directory) }

        let repository = ClipRepository(databaseURL: directory.appendingPathComponent("clips.sqlite"))
        let old = ClipItem(
            id: "old",
            createdAt: Date(timeIntervalSince1970: 1),
            type: .media,
            plainText: nil,
            payloadPath: directory.appendingPathComponent("old.png").path,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "old-hash",
            origin: .original,
            derivedFromClipID: nil
        )
        let new = ClipItem(
            id: "new",
            createdAt: Date(timeIntervalSince1970: 2),
            type: .text,
            plainText: "new",
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "new-hash",
            origin: .original,
            derivedFromClipID: nil
        )

        try repository.insert(old)
        try repository.insert(new)

        XCTAssertEqual(repository.payloadPathsBeyondLimit(maxCount: 1), [old.payloadPath!])
        repository.prune(maxCount: 1)

        let clips = repository.fetch(limit: 10)
        XCTAssertEqual(clips.map(\.id), ["new"])
    }

    func testMediaFilterIncludesCopiedMediaFiles() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: directory) }

        let repository = ClipRepository(databaseURL: directory.appendingPathComponent("clips.sqlite"))
        let mediaPayloadClip = ClipItem(
            id: "media-payload",
            createdAt: Date(timeIntervalSince1970: 1),
            type: .media,
            plainText: nil,
            payloadPath: directory.appendingPathComponent("payload.png").path,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "payload-hash",
            origin: .original,
            derivedFromClipID: nil
        )
        let mediaFileClip = ClipItem(
            id: "media-file",
            createdAt: Date(timeIntervalSince1970: 2),
            type: .media,
            plainText: directory.appendingPathComponent("named-photo.png").path,
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "file-hash",
            origin: .original,
            derivedFromClipID: nil
        )
        let documentClip = ClipItem(
            id: "document",
            createdAt: Date(timeIntervalSince1970: 3),
            type: .document,
            plainText: directory.appendingPathComponent("notes.pdf").path,
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "document-hash",
            origin: .original,
            derivedFromClipID: nil
        )

        try repository.insert(mediaPayloadClip)
        try repository.insert(mediaFileClip)
        try repository.insert(documentClip)

        let clips = repository.fetch(filter: .media, limit: 10)
        XCTAssertEqual(clips.map(\.id), ["media-file", "media-payload"])
    }

    func testDeleteMatchingFilterRemovesOnlyCurrentCategory() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: directory) }

        let repository = ClipRepository(databaseURL: directory.appendingPathComponent("clips.sqlite"))
        let textClip = ClipItem(
            id: "text",
            createdAt: Date(timeIntervalSince1970: 1),
            type: .text,
            plainText: "hello",
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "text-hash",
            origin: .original,
            derivedFromClipID: nil
        )
        let mediaClip = ClipItem(
            id: "media",
            createdAt: Date(timeIntervalSince1970: 2),
            type: .media,
            plainText: nil,
            payloadPath: directory.appendingPathComponent("image.png").path,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "media-hash",
            origin: .original,
            derivedFromClipID: nil
        )
        let documentClip = ClipItem(
            id: "document",
            createdAt: Date(timeIntervalSince1970: 3),
            type: .document,
            plainText: directory.appendingPathComponent("notes.pdf").path,
            payloadPath: nil,
            sourceApp: "Tests",
            isPinned: false,
            contentHash: "document-hash",
            origin: .original,
            derivedFromClipID: nil
        )

        try repository.insert(textClip)
        try repository.insert(mediaClip)
        try repository.insert(documentClip)

        XCTAssertEqual(repository.payloadPaths(matching: .media), [mediaClip.payloadPath!])
        repository.delete(matching: .media)

        XCTAssertEqual(repository.fetch(limit: 10).map(\.id), ["document", "text"])
    }
}

File: windows/Clipwell.Win/tests/Clipwell.Core.Tests/ClipRepositoryTests.cs
using Clipwell.Core.Data;
using Clipwell.Core.Models;
using Xunit;

namespace Clipwell.Core.Tests;

public sealed class ClipRepositoryTests : IDisposable
{
    private readonly string _tempDir;
    private readonly ClipRepository _repository;

    public ClipRepositoryTests()
    {
        _tempDir = Directory.CreateTempSubdirectory("clipwell-tests-").FullName;
        _repository = new ClipRepository(Path.Combine(_tempDir, "clips.sqlite"));
    }

    public void Dispose()
    {
        _repository.Dispose();
        Directory.Delete(_tempDir, recursive: true);
    }

    private static ClipItem MakeClip(
        string text,
        ClipType type = ClipType.Text,
        bool isPinned = false,
        DateTimeOffset? createdAt = null) =>
        new(
            Id: Guid.NewGuid().ToString(),
            CreatedAt: createdAt ?? DateTimeOffset.UtcNow,
            Type: type,
            PlainText: text,
            PayloadPath: null,
            SourceApp: "TestApp",
            IsPinned: isPinned,
            ContentHash: Clipwell.Core.Services.ContentHasher.Hash(text),
            Origin: ClipOrigin.Original,
            DerivedFromClipId: null);

    [Fact]
    public void InsertAndFetchRoundTrips()
    {
        var clip = MakeClip("hello world");
        _repository.Insert(clip);

        var fetched = Assert.Single(_repository.Fetch());
        Assert.Equal(clip.Id, fetched.Id);
        Assert.Equal("hello world", fetched.PlainText);
        Assert.Equal(ClipType.Text, fetched.Type);
        Assert.Equal(ClipOrigin.Original, fetched.Origin);
        Assert.Equal("TestApp", fetched.SourceApp);
        Assert.Equal(clip.ContentHash, fetched.ContentHash);
        Assert.Equal(clip.CreatedAt.ToUnixTimeMilliseconds(), fetched.CreatedAt.ToUnixTimeMilliseconds());
    }

    [Fact]
    public void SearchMatchesPlainTextSubstring()
    {
        _repository.Insert(MakeClip("alpha beta"));
        _repository.Insert(MakeClip("gamma delta"));

        var results = _repository.Fetch(search: "beta");
        Assert.Equal("alpha beta", Assert.Single(results).PlainText);
    }

    [Fact]
    public void FilterSelectsTextLikeTypes()
    {
        _repository.Insert(MakeClip("plain"));
        _repository.Insert(MakeClip("<b>rich</b>", ClipType.Html));
        _repository.Insert(MakeClip("/tmp/movie.mp4", ClipType.Media));
        _repository.Insert(MakeClip("/tmp/file.pdf", ClipType.Document));

        Assert.Equal(2, _repository.Fetch(filter: ClipFilter.Text).Count);
        Assert.Single(_repository.Fetch(filter: ClipFilter.Media));
        Assert.Single(_repository.Fetch(filter: ClipFilter.Document));
        Assert.Equal(4, _repository.Fetch(filter: ClipFilter.All).Count);
    }

    [Fact]
    public void FetchOrdersPinnedFirstThenNewest()
    {
        var baseTime = DateTimeOffset.UtcNow;
        _repository.Insert(MakeClip("oldest", createdAt: baseTime.AddMinutes(-3)));
        _repository.Insert(MakeClip("pinned-old", isPinned: true, createdAt: baseTime.AddMinutes(-2)));
        _repository.Insert(MakeClip("newest", createdAt: baseTime));

        var results = _repository.Fetch();
        Assert.Equal(["pinned-old", "newest", "oldest"], results.Select(clip => clip.PlainText));
    }

    [Fact]
    public void PruneKeepsPinnedAndNewest()
    {
        var baseTime = DateTimeOffset.UtcNow;
        _repository.Insert(MakeClip("pinned", isPinned: true, createdAt: baseTime.AddMinutes(-10)));
        for (var i = 0; i < 5; i++)
        {
            _repository.Insert(MakeClip($"clip-{i}", createdAt: baseTime.AddMinutes(-5 + i)));
        }

        _repository.Prune(maxCount: 2);

        var remaining = _repository.Fetch();
        Assert.Equal(3, remaining.Count);
        Assert.Contains(remaining, clip => clip.PlainText == "pinned");
        Assert.Contains(remaining, clip => clip.PlainText == "clip-4");
        Assert.Contains(remaining, clip => clip.PlainText == "clip-3");
    }

    [Fact]
    public void LatestClipReturnsNewestForDedup()
    {
        _repository.Insert(MakeClip("first", createdAt: DateTimeOffset.UtcNow.AddMinutes(-1)));
        var latest = MakeClip("second");
        _repository.Insert(latest);

        Assert.Equal(latest.ContentHash, _repository.LatestClip()?.ContentHash);
    }

    [Fact]
    public void SetPinnedUpdatesRow()
    {
        var clip = MakeClip("pin me");
        _repository.Insert(clip);
        _repository.SetPinned(clip.Id, true);

        Assert.True(Assert.Single(_repository.Fetch()).IsPinned);
    }

    [Fact]
    public void DeleteAndClearRemoveRows()
    {
        var clip = MakeClip("bye");
        _repository.Insert(clip);
        _repository.Delete(clip.Id);
        Assert.Empty(_repository.Fetch());

        _repository.Insert(MakeClip("a"));
        _repository.Insert(MakeClip("b"));
        _repository.Clear();
        Assert.Empty(_repository.Fetch());
    }

    [Fact]
    public void PayloadPathsBeyondLimitReturnsOnlyPrunablePayloads()
    {
        var baseTime = DateTimeOffset.UtcNow;
        var old = MakeClip("old", ClipType.Media, createdAt: baseTime.AddMinutes(-2)) with { PayloadPath = "/tmp/old.png" };
        var fresh = MakeClip("new", ClipType.Media, createdAt: baseTime) with { PayloadPath = "/tmp/new.png" };
        _repository.Insert(old);
        _repository.Insert(fresh);

        Assert.Equal(["/tmp/old.png"], _repository.PayloadPathsBeyondLimit(maxCount: 1));
    }
}
